draco/asp/soft.lp

```diff
@@ -30,18 +30,67 @@ preference(same_field,F) :-
     entity(mark,_,M),
     { entity(encoding,M,E): attribute((encoding,field),E,F) } = 2.
 
-% @soft(type_linear) linear scale
-preference(type_linear,E) :-
+% @soft(same_field_grt3) Prefer not to use the same field three or more times for the same mark.
+preference(same_field_grt3,F) :-
+    entity(field,_,F),
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,field),E,F) } >= 3.
+
+% @soft(count_grt1) Prefer not to use count more than once.
+preference(count_grt1,M) :-
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,aggregate),E,count) } > 1.
+
+% @soft(shape_high_cardinality) Shape channel should not have too high cardinality.
+preference(shape_high_cardinality,E) :-
+    attribute((encoding,channel),E,shape),
+    attribute(encoding_cardinality,E,N),
+    N > 5.
+
+% @soft(number_categorical) Numbers should not be nominal.
+preference(number_categorical,F) :-
+    attribute((field,type),F,number),
+    attribute(field_scale,F,categorical).
+
+% @soft(bin_number_low_cardinality) Binned number field should not have too low cardinality.
+preference(bin_number_low_cardinality,E) :-
+    attribute((field,type),F,number),
```

Comment:
datetime might also work as quantitative data type
```suggestion
    attribute((field,type),F,(number;datetime)),
```

Reply:
There is a hard constraint that binning can only be added to number/datatime. And also there are two soft constraints bin_low and bin_high, so I wonder if I should keep this one instead.

Reply:
but soft constraint is doing a different thing from hard constraint. I think you still need to include the datatime for the purpose of counting the preference.

Reply:
I typed a word wrongly lol. I meant maybe we should *remove* this soft constraint since there are hard and other similar soft

Reply:
but `bin_low/high` is about the number of bins; this one is about field cardinality